cmake/DAGMC_macros.cmake

```diff
@@ -0,0 +1,271 @@
+    set(CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS)
+    set(CMAKE_SHARED_LIBRARY_LINK_Fortran_FLAGS)
+    set(CMAKE_EXE_LINK_DYNAMIC_C_FLAGS)
+    set(CMAKE_EXE_LINK_DYNAMIC_CXX_FLAGS)
+    set(CMAKE_EXE_LINK_DYNAMIC_Fortran_FLAGS)
+  else ()
+    message(STATUS "Building shared executables")
+    set(BUILD_SHARED_EXE ON)
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".so")
+  endif ()
+
+  if (BUILD_RPATH)
+    set(INSTALL_RPATH_DIRS "${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}")
+    message(STATUS "INSTALL_RPATH_DIRS: ${INSTALL_RPATH_DIRS}")
+  endif ()
+
+  message(STATUS "CMAKE_C_FLAGS: ${CMAKE_C_FLAGS}")
+  message(STATUS "CMAKE_CXX_FLAGS: ${CMAKE_CXX_FLAGS}")
+  message(STATUS "CMAKE_Fortran_FLAGS: ${CMAKE_Fortran_FLAGS}")
+  message(STATUS "CMAKE_C_FLAGS_${CMAKE_BUILD_TYPE_UPPER}: ${CMAKE_C_FLAGS_${CMAKE_BUILD_TYPE_UPPER}}")
+  message(STATUS "CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE_UPPER}: ${CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE_UPPER}}")
+  message(STATUS "CMAKE_Fortran_FLAGS_${CMAKE_BUILD_TYPE_UPPER}: ${CMAKE_Fortran_FLAGS_${CMAKE_BUILD_TYPE_UPPER}}")
+  message(STATUS "CMAKE_C_IMPLICIT_LINK_LIBRARIES: ${CMAKE_C_IMPLICIT_LINK_LIBRARIES}")
+  message(STATUS "CMAKE_CXX_IMPLICIT_LINK_LIBRARIES: ${CMAKE_CXX_IMPLICIT_LINK_LIBRARIES}")
+  message(STATUS "CMAKE_Fortran_IMPLICIT_LINK_LIBRARIES: ${CMAKE_Fortran_IMPLICIT_LINK_LIBRARIES}")
+  message(STATUS "CMAKE_EXE_LINKER_FLAGS: ${CMAKE_EXE_LINKER_FLAGS}")
+endmacro ()
+
+# Figure out what LINK_LIBS_SHARED and LINK_LIBS_STATIC should be based on the
+# values of LINK_LIBS and LINK_LIBS_EXTERN_NAMES
+macro (dagmc_get_link_libs)
+  set(LINK_LIBS_SHARED)
+  set(LINK_LIBS_STATIC)
+
+  foreach (extern_name IN LISTS LINK_LIBS_EXTERN_NAMES)
+    list(APPEND LINK_LIBS_SHARED ${${extern_name}_SHARED})
+    list(APPEND LINK_LIBS_STATIC ${${extern_name}_STATIC})
+  endforeach ()
+
+  foreach (link_lib IN LISTS LINK_LIBS)
+    list(FIND DAGMC_LIBRARY_LIST ${link_lib} index)
+    if (index STREQUAL "-1")
+      list(APPEND LINK_LIBS_SHARED ${link_lib})
+      list(APPEND LINK_LIBS_STATIC ${link_lib})
+    else ()
+      list(APPEND LINK_LIBS_SHARED ${link_lib}-shared)
+      list(APPEND LINK_LIBS_STATIC ${link_lib}-static)
+    endif ()
+  endforeach ()
+endmacro ()
+
+# Setup the configuration file and install
+macro (dagmc_make_configure_file)
+  message("")
+
+  message(STATUS "DAGMC cmake config file: ${CMAKE_INSTALL_PREFIX}/${INSTALL_LIB_DIR}/cmake/DAGMCConfig.cmake")
+  configure_file(cmake/DAGMCConfig.cmake.in DAGMCConfig.cmake @ONLY)
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/DAGMCConfig.cmake DESTINATION ${INSTALL_LIB_DIR}/cmake/)
+endmacro ()
+
+# To use the dagmc_install macros, the following lists must be defined:
+#   SRC_FILES: source files
+#   PUB_HEADERS: public header files
+#   LINK_LIBS: e.g. dagmc, pyne_dagmc, uwuw, lapack, gfortran
+#   LINK_LIBS_EXTERN_NAMES: e.g. MOAB_LIBRARIES
+
+# Install a library in both shared and static mode
+macro (dagmc_install_library lib_name)
```

Comment:
can you clarify what you want to see here

Reply:
I was thinking that this macro might accept other variables like `LINK_LIBS`, `SRC_FILES`, and `PUB_HEADERS`. This way we can locally name variables like `DAGTALLY_LINK_LIBS`, `DAGTALLY_SRC_FILES`, and `DAGTALLY_PUB_HEADERS` instead of overwriting the same variable names again and again. I think it could lead to a lot of confusion if they aren't updated properly somewhere in the config.

```
dagmc_install_library(dagtally DAGTALLY_LINK_LIBS DAGTALLY_SRC_FILES DAGTALLY_PUB_HEADERS)

dagmc_link_libs(DAGMC_LINK_LIBS)
```

Reply:
OR

you could maybe leave the signature the way it is and expect that variables `<lib_name>_LINK_LIBS`, `<lib_name>_SRC_FILES`, and `<lib_name>_PUB_HEADERS` are defined.

Reply:
I'm not sure how difficult/possible either of those alternatives are. Just spit-balling here.

Reply:
I did this in a new branch... not really sure if I like it though. https://github.com/ljacobson64/DAGMC/tree/macro_args

Reply:
This also seems to make it take a lot longer to run cmake as well, and I don't really know why